### ISSUE TITLE
Proposal: reorganize material to remove 'publish' and introduce 'provenance'

### ISSUE THREAD

**Issue by gvwilson** [merely-useful/merely-useful.github.io#421]
## Proposal

1. Reorganize material as described in https://github.com/merely-useful/merely-useful.github.io/pull/415.
2. Do *not* introduce explicit parts in the ToC: they would be of disproportionate sizes (and annoying to implement).
3. Move data documentation material from the `project` chapter to the new `provenance` chapter.
4. Move the boilerplate material and project structure material to the `teams` chapter.
5. [late addition] Merge all outstanding changes and send to publishers as-is with a note saying that we're reorganizing some material.

## Background

See discussion in https://github.com/merely-useful/merely-useful.github.io/pull/415.

## Pros and Cons

**Pro:**
-   The `publish` chapter is a weak way to end the book.

**Con:**
-   It's a bit late in the day to be doing this.

## Alternatives

1. Leave things the way they are: this will produce a weaker book.
2. Cut all the `publish` material: but some of it is very useful.

**Comment by gvwilson** [comment]
As well as voting up or down, please add a comment to let us know if you'll be able to spend 2-3 hours between now and July 24 looking over the rearrangement.

**Comment by DamienIrving** [comment]
Thanks for starting this proposal, @gvwilson. The order in which all the content presented is probably the last tricky thing we need to sort out.

In addition to the discussion in #415 (which makes the case for the new `provenance.md` chapter), the other thread to bring in is #420 (about the project content).

Here's some thoughts on the ordering (comments welcome, I'm still thinking through this and my thoughts change often!)...

First we have the core of the book, which involves some foundational skills (we can't do anything until the reader is comfortable at the command line) followed by the ordered sequence of events involved in developing the Zipf's Law software. In other words, the order of these chapters is set by the narrative and they can't really be moved around.
```
  - py-rse/bash-basics.Rmd
  - py-rse/bash-advanced.Rmd
  - py-rse/scripting.Rmd
  - py-rse/git-cmdline.Rmd
  - py-rse/git-advanced.Rmd
  - py-rse/automate.Rmd
  - py-rse/config.Rmd
  - py-rse/errors.Rmd
  - py-rse/testing.Rmd
  - py-rse/provenance.Rmd   
  - py-rse/packaging.Rmd 
``` 

We then have the content relating to project organisation (`py-rse/project.Rmd`), code style (`py-rse/style.Rmd`) and working in teams (`py-rse/teams.Rmd`), which is important stuff, but doesn't have such a set place in the sequence of events:
* From the moment we start writing Python code in the scripting chapter all the way through to our last line of new Python code in the packaging chapter, we adopt the principles of `style.Rmd`
* From the initial files and directory structure in the bash basics chapter all the way through to the packaging chapter, we add files and directories to our project in a manner consistent with the principles of `project.Rmd`
* The concepts introduced in `teams.Rmd` relating to working effectively with others might be applicable from the very beginning (e.g. Amira and Sami working together on the Zipf's Law analysis) or not until the software is packaged and released into the wild and the community gets involved... every reader's experience / project will be different.

In other words, these chapters are kind of floating content that is less part of the narrative and more a set of reference material/principles that we should be referring back to continuously throughout the narrative. I'm thinking they probably therefore need their own section of the book, perhaps called "bonus content" or "reference section" or "guiding principles"?

**Comment by gvwilson** [comment]
1. I would like to put `teams` immediately after `git-advanced`: the `git-advanced` chapter shows people how to collaborate with others on GitHub, so (a) that feels like a natural place to introduce the rules for productive collaboration and (b) if we put it at the end, it will feel like an afterthought.
2. I think `project` should come before `packaging` - you can have a sensibly-organized project that isn't a package, and `project` feels less intimidating to me than `packaging`. (I'm still iffy about `virtualenv`, but I realize we need to introduce it if we're going to have people build packages.)

**Comment by cwickham** [comment]
Agree with Greg that we should introduce `teams` as soon as possible, after `git-advanced` seems good.

Also agree that `project` could fit before packaging.  

I can spend some time reviewing reorganization next week.

**Comment by k8hertweck** [comment]
I also agree with the order described up to this point in the discussion thread. If we move `teams` up that far in the book, it may be possible to streamline the content a bit more (as it's currently a bit bulky). I can commit time to do this or other sections over the next few weeks.

**Comment by lwjohnst86** [comment]
I would have time between July 20 to 24 to work a bit on it.

**Comment by gvwilson** [comment]
This sounds like a plan, so may I ask:

-   @lwjohnst86 to do the rearranging ASAP (starting July 20 if possible)
-   @cwickham to do the first review and flag anything that needs attention that @lwjohnst86 hasn't already flagged

If this can be completed by Monday July 27, I can do another pass for grammar and continuity next week. Meanwhile, could @k8hertweck and @DamienIrving please go through the later chapters this week as well and add exercises and/or solutions?

**Comment by lwjohnst86** [comment]
Sounds good. I can get to it tomorrow!

**Comment by k8hertweck** [comment]
Yup, I'll work on exercises. I'll start a PR for each chapter I'm targeting so @DamienIrving will be aware.

**Comment by cwickham** [comment]
I'm on board too.  @lwjohnst86, ping me when your ready for me to take a look.

**Comment by DamienIrving** [comment]
Just to clarify, it sounds like we're pretty settled on the following order (as it appears in `_py-rse.yml`):

```
  - py-rse-index.Rmd
  - py-rse/bash-basics.Rmd
  - py-rse/bash-advanced.Rmd
  - py-rse/scripting.Rmd
  - py-rse/git-cmdline.Rmd
  - py-rse/git-advanced.Rmd
  - py-rse/teams.Rmd
  - py-rse/automate.Rmd
  - py-rse/config.Rmd
  - py-rse/errors.Rmd
  - py-rse/testing.Rmd
  - py-rse/provenance.Rmd
  - py-rse/project.Rmd   
  - py-rse/packaging.Rmd 
  - py-rse/finale.Rmd
```
We haven't discussed a spot for `py-rse/style.Rmd`... maybe @lwjohnst86 can propose a spot in his forthcoming PR for the reorganization? 

In terms of the content for the new provenance chapter (much of which will be repurposed from the old publishing chapter and then edited according to the comments made in #415), I'd started working on that last week so I've thrown what I've done so far into a draft PR at #422. I'll try and get as much of that done as possible before the meeting this week so we can have a meaningful discussion about what stays and what goes in that chapter.

**Comment by lwjohnst86** [comment]
@DamienIrving  I didn't make any change to provenance aside from including it in the book outline.

**Comment by DamienIrving** [comment]
Ok, so #423 re-orders the chapters and #422 massages the old content from `publish.Rmd` into the new `provenance.Rmd`.
